pricegraph/src/graph/bellman_ford.rs

```diff
@@ -2,40 +2,170 @@
+        let PredecessorStore(predecessor, _) = self;
+        predecessor[node_index] = updated_predecessor;
+    }
+}
+
+/// Structure that can be used to derive the shorthest path from a source to any
+/// reachable destination in the graph.
+pub struct ShortestPathGraph<G: GraphBase + Data> {
+    graph: G,
+    predecessor_store: PredecessorStore<G>,
+    source: G::NodeId,
+}
+
+impl<G> ShortestPathGraph<G>
+where
+    G: IntoEdges + NodeIndexable,
+{
+    /// Returns the current distance of a node from the source.
+    fn distance(&self, node: G::NodeId) -> &G::EdgeWeight {
+        self.predecessor_store.distance(self.graph.to_index(node))
+    }
+
+    /// Updates the distance of a node from the source.
+    fn update_distance(&mut self, node: G::NodeId, updated_distance: G::EdgeWeight) {
+        self.predecessor_store
+            .update_distance(self.graph.to_index(node), updated_distance);
+    }
+
+    /// Updates the predecessor of a node.
+    fn update_predecessor(&mut self, node: G::NodeId, updated_predecessor: Option<G::NodeId>) {
+        self.predecessor_store
+            .update_predecessor(self.graph.to_index(node), updated_predecessor);
+    }
+
+    /// Returns shortest path from source to destination node, if a path exists.
+    pub fn path_to(&self, dest: G::NodeId) -> Option<Path<G::NodeId>> {
+        let mut path;
+        let mut current = dest;
+        let PredecessorStore(predecessor, _) = &self.predecessor_store;
+        path = Vec::with_capacity(predecessor.len());
+        while current != self.source {
+            path.push(current);
+            current = predecessor[self.graph.to_index(current)]?;
+        }
+        path.push(self.source);
+
+        // NOTE: `path` is in reverse order, since it was built by walking the path
+        // backwards, so reverse it and done!
+        path.reverse();
+        Some(Path(path))
+    }
+
+    /// Lists all nodes that can be reached from the source.
+    pub fn connected_nodes(&self) -> Vec<G::NodeId> {
+        let PredecessorStore(predecessor, _) = &self.predecessor_store;
+        let mut node_indices: Vec<_> = predecessor
+            .iter()
+            .enumerate()
+            .filter_map(|(i, &pre)| pre.map(|_| i))
+            .collect();
+        // if the source was not connected to anything, no node would have it as its
+        // predecessor
+        node_indices.push(self.graph.to_index(self.source));
+        node_indices.sort();
+        node_indices.dedup();
+        node_indices
+            .into_iter()
+            .map(|i| self.graph.from_index(i))
+            .collect()
+    }
+}
+
+impl<G> ShortestPathGraph<G>
+where
+    G: IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    /// Initializes a shortest path graph that will be later built with the
+    /// Bellman-Ford algorithm.
+    fn init(g: G, source: G::NodeId) -> ShortestPathGraph<G> {
```

Comment:
Could call it `new` which is more common in Rust but if init has a specific meaning different from new then that is fine too.

In this function signature and body you can replace `ShortestPathGraph<G>` and `ShortestPathGraph` with `Self` for less repetition.